### PR TITLE
add Promise\wrap

### PR DESCRIPTION
**Pull request by prolic**

### CODE REVIEW

**Comment by bwoebi** [comment]
Hm, we had quite a similar function in Amp v1: `\Amp\pipe`: https://github.com/amphp/amp/blob/1.x/lib/functions.php#L597-L639

We removed it in v2 in favor of generators; i.e.:

```php
\Amp\call(function() use ($promise) {
    return doSomeTransformation(yield $promise);
});
```

So I guess we won't accept this either, for the same reason.

**Comment by prolic** [comment]
@bwoebi thanks, i like your suggestion even more

**Comment by kelunik** [comment]
@bwoebi This is different from pipe, as it also gets executed on errors. I'd probably merge it.

**Comment by bwoebi** [comment]
@kelunik Sure, but you can easily try/catch that in a coroutine, if you wish error handling - hence "similar".

File: lib/functions.php [hunk 1]
```diff
@@ -441,6 +441,34 @@ function some(array $promises, int $required = 1): Promise
...
     }
+
+    /**
+     * Wraps a promise into another promise, altering the exception or result
+     *
+     * @param Promise $promise
+     * @param callable $callback
+     * @return Promise
+     */
+    function wrap(Promise $promise, callable $callback): Promise
```
Comment (kelunik) [inline lib/functions.php]: Could you please adjust this PR to accept a ReactPHP promise like all other functions?

Comment (prolic) [reply lib/functions.php]: done

Comment (kelunik) [reply lib/functions.php]: A passed ReactPHP promise needs to be adapted before `onResolve()` can be called.

Comment (prolic) [reply lib/functions.php]: ah totally missed that, my bad :)

**Pull request merged**